File: leetcode/studyplan_algorithm1/two_pointers/two_sum_ii/main.rs
// TODO: Update Comment
// Time Complexity: O(n^2)
// Space Complexity: O(1)
fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
    let mut res: Vec<i32> = Vec::new();

    for i in 0..numbers.len() {
        if res.len() == 2 {
            break;
        }
        for j in i + 1..numbers.len() {
            if numbers[i] + numbers[j] == target {
                res.push(i as i32 + 1);
                res.push(j as i32 + 1);
            }
        }
    }

    res
}

fn main() {
    println!("{:?}", two_sum(vec![2, 7, 11, 15], 9));
}

#[cfg(test)]
mod tests {
    use crate::two_sum;

    #[test]
    fn test_positive_vector() {
        assert_eq!(two_sum(vec![2, 7, 11, 15], 9), vec![1, 2]);
        assert_eq!(two_sum(vec![2, 3, 4], 6), vec![1, 3]);
    }

    #[test]
    fn test_negative_vector() {
        assert_eq!(two_sum(vec![-1, 0], -1), vec![1, 2]);
    }
}

File: leetcode/studyplan_algorithm1/two_pointers/two_sum_ii/main.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> twoSum(vector<int>& numbers, int target) {
  vector<int> res;
  
  // O(n^2)
  for (int i = 0; i < numbers.size(); i++) 
    for (int j = i + 1; j < numbers.size(); j++) 
      if (numbers[i] + numbers[j] == target) {
        res.push_back(i+1);
        res.push_back(j+1);
      }

  return res;
}

vector<int> twoSumFast(vector<int>& numbers, int target) {
  vector<int> res;

  int left = 0, right = numbers.size() - 1;

  // O(n)
  while (left < right) {
    if (numbers[left] + numbers[right] == target) {
      res.push_back(left+1);
      res.push_back(right+1);
      break;
    }
    else if (numbers[left] + numbers[right] < target) left++;
    else right--;
  }
  
  return res;
}

int main() {
  vector<int> tc1{2,7,11,15,16,19,234234,235543,534543};
  
  // vector<int> result = twoSum(tc1, 768777);
  vector<int> result = twoSumFast(tc1, 768777);

  for (int i = 0; i < result.size(); i++)
    cout << result[i] << ' ';
  cout << "\n";
}


